astropy/io/fits/scripts/fitscheck.py

```diff
@@ -130,39 +125,41 @@ def verify_checksums(filename):
     Prints a message if any HDU in `filename` has a bad checksum or datasum.
     """
 
-    errors = 0
-    try:
-        hdulist = fits.open(filename, checksum=OPTIONS.checksum_kind)
-    except UserWarning as w:
-        remainder = '.. ' + ' '.join(str(w).split(' ')[1:]).strip()
-        # if "Checksum" in str(w) or "Datasum" in str(w):
-        log.warning('BAD {!r} {}'.format(filename, remainder))
-        return 1
-    if not OPTIONS.ignore_missing:
-        for i, hdu in enumerate(hdulist):
-            if not hdu._checksum:
-                log.warning('MISSING {!r} .. Checksum not found '
-                            'in HDU #{}'.format(filename, i))
-                return 1
-            if not hdu._datasum:
-                log.warning('MISSING {!r} .. Datasum not found '
-                            'in HDU #{}'.format(filename, i))
-                return 1
-    if not errors:
-        log.info('OK {!r}'.format(filename))
-    return errors
+    with catch_warnings() as wlist:
+        with fits.open(filename, checksum=OPTIONS.checksum_kind) as hdulist:
+            for i, hdu in enumerate(hdulist):
+                # looping on HDUs is needed to read them and verify the
+                # checksums
+                if not OPTIONS.ignore_missing:
+                    if not hdu._checksum:
+                        log.warning('MISSING {!r} .. Checksum not found '
+                                    'in HDU #{}'.format(filename, i))
+                        return 1
+                    if not hdu._datasum:
+                        log.warning('MISSING {!r} .. Datasum not found '
+                                    'in HDU #{}'.format(filename, i))
+                        return 1
+
+    for w in wlist:
+        if str(w.message).startswith(('Checksum verification failed',
```

Comment:
This would maybe be a bit cleaner if we used a new warning subclass for all FITS checksum problems instead of just `AstropyUserWarning`.